In [1]:
import scipy
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
accident.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

In [6]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [7]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'] , dayfirst = True, errors = 'coerce')

In [8]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious           NaT  51.517715   
3       200701BS66128           Serious           NaT  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight           NaT  57.374005   
660675  201091NM01881            Slight           NaT  57.232273   
660676  201091NM01935            Slight           NaT  57.585044   
660677  201091NM01964           Serious           NaT  57.214898   
660678  201091NM02142           Serious           NaT  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [9]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [10]:
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                          object
Urban_or_Rural_Area              category
Weather_Conditions                 object
Vehicle_Type                     category
dtype: object

In [11]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [12]:
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode())

In [13]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode())
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode())

In [16]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])

In [28]:
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted weather conditions')

TypeError: Cannot setitem on a Categorical with a new category (unaccounted), set the categories first

In [31]:
accident.isnull().sum()

Index                           0
Accident_Severity               0
Accident Date              395672
Latitude                       25
Light_Conditions                0
District Area                   0
Longitude                      26
Number_of_Casualties            0
Number_of_Vehicles              0
Road_Surface_Conditions         0
Road_Type                    4520
Urban_or_Rural_Area             0
Weather_Conditions          14128
Vehicle_Type                    0
Year                       395672
Month                      395672
Day                        395672
DayOfWeek                  395672
dtype: int64

In [32]:
accident['Urban_or_Rural_Area'].mode()

0    Urban
Name: Urban_or_Rural_Area, dtype: category
Categories (3, object): ['Rural', 'Unallocated', 'Urban']

In [33]:
accident[accident['District Area'] == 'Kensington and Chelsea'].count()

Index                      2799
Accident_Severity          2799
Accident Date              1130
Latitude                   2799
Light_Conditions           2799
District Area              2799
Longitude                  2799
Number_of_Casualties       2799
Number_of_Vehicles         2799
Road_Surface_Conditions    2799
Road_Type                  2796
Urban_or_Rural_Area        2799
Weather_Conditions         2797
Vehicle_Type               2799
Year                       1130
Month                      1130
Day                        1130
DayOfWeek                  1130
dtype: int64

In [34]:
accident_agg = accident.groupby(['District Area']).size()

In [35]:
accident_agg

District Area
Aberdeen City    1323
Aberdeenshire    1930
Adur              619
Allerdale        1128
Alnwick           232
                 ... 
Wychavon         1361
Wycombe          1729
Wyre             1238
Wyre Forest       969
York             1897
Length: 422, dtype: int64

In [36]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek 

In [37]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious           NaT  51.517715   
3       200701BS66128           Serious           NaT  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight           NaT  57.374005   
660675  201091NM01881            Slight           NaT  57.232273   
660676  201091NM01935            Slight           NaT  57.585044   
660677  201091NM01964           Serious           NaT  57.214898   
660678  201091NM02142           Serious           NaT  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type    Year  Month  Day  DayOfWeek  
0                                        Car  2019.0    6.0  5.0        2.0  
1                                        Car  2019.0    7.0  2.0        1.0  
2                      Taxi/Private hire car     NaN    NaN  NaN        NaN  
3       Bus or coach (17 or more pass seats)     NaN    NaN  NaN        NaN  
4                              Other vehicle  2019.0    9.0  3.0        1.0  
...                                      ...     ...

In [57]:
acci_2019 = accident[accident['Year'] == 2019]
acci_2021 = accident[accident['Year'] == 2021]

<h1>UniVariate</h1>
<h2>Question 1</h2>
<h3>How many accidents happen in 2020</h3>

In [51]:
acci_2020 = accident[accident['Year'] == 2020]
acci_2020

Index Accident_Severity Accident Date   Latitude  \
182120  200801BS70001           Serious    2020-01-02  51.491521   
182121  200801BS70002            Slight    2020-01-06  51.484909   
182122  200801BS70003           Serious    2020-01-12  51.509965   
182123  200801BS70004            Slight    2020-01-05  51.490631   
182124  200801BS70005            Slight    2020-01-03  51.509712   
...               ...               ...           ...        ...   
352694       2.01E+12            Slight    2020-11-01  55.267100   
352695       2.01E+12            Slight    2020-11-02  54.985218   
352696       2.01E+12            Slight    2020-11-06  55.126057   
352697       2.01E+12           Serious    2020-11-04  55.243971   
352702       2.01E+12            Slight    2020-12-06  54.983911   

              Light_Conditions           District Area  Longitude  \
182120   Darkness - lights lit  Kensington and Chelsea  -0.179406   
182121   Darkness - lights lit  Kensington and Chelsea  -0.176356   
182122                Daylight  Kensington and Chelsea  -0.197405   
182123   Darkness - lights lit  Kensington and Chelsea  -0.185779   
182124   Darkness - lights lit  Kensington and Chelsea  -0.192659   
...                        ...                     ...        ...   
352694  Darkness - no lighting   Dumfries and Galloway  -2.959324   
352695                Daylight   Dumfries and Galloway  -3.278743   
352696  Darkness - no lighting   Dumfries and Galloway  -3.344404   
352697  Darkness - no lighting   Dumfries and Galloway  -3.461163   
352702                Daylight   Dumfries and Galloway  -3.195094   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
182120                     1                   2                     Dry   
182121                     1                   2             Wet or damp   
182122                     1                   1                     Dry   
182123                     1                   2                     Dry   
182124                     3                   1                     Dry   
...                      ...                 ...                     ...   
352694                     1                   1            Frost or ice   
352695                     1                   1             Wet or damp   
352696                     1                   1             Wet or damp   
352697                     2                   2                     Dry   
352702                     1                   1            Frost or ice   

                 Road_Type Urban_or_Rural_Area  Weather_Conditions  \
182120  Single carriageway               Urban  Fine no high winds   
182121  Single carriageway               Urban  Fine no high winds   
182122  Single carriageway               Urban  Fine no high winds   
182123  Single carriageway               Urban  Fine no high winds   
182124  Single carriageway               Urban  Fine no high winds   
...                    ...                 ...                 ...   
352694  Single carriageway               Rural  Fine no high winds   
352695  Single carriageway               Rural  Fine no high winds   
352696  Single carriageway               Rural  Fine no high winds   
352697  Single carriageway               Rural  Fine no high winds   
352702  Single carriageway               Rural  Fine no high winds   

                      Vehicle_Type    Year  Month   Day  DayOfWeek  
182120                         Car  2020.0    1.0   2.0        3.0  
182121       Motorcycle over 500cc  2020.0    1.0   6.0        0.0  
182122                         Car  2020.0    1.0  12.0        6.0  
182123                         Car  2020.0    1.0   5.0        6.0  
182124                         Car  2020.0    1.0   3.0        4.0  
...                            ...     ...    ...   ...        ...  
352694                         Car  2020.0   11.0   1.0        6.0  
352695  Motorcycle 125cc and under  2020.0   11.0   2.0        0

In [39]:
acci_2020['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       51605
Darkness - lights lit          13467
Darkness - no lighting          4085
Darkness - lighting unknown      717
Darkness - lights unlit          289
Name: count, dtype: int64

In [40]:
acci_2020['Light_Conditions'].value_counts().sum()

np.int64(70163)

<h1>Insight 1 =There are 70,163 accidents happen in 2020 and There are 73.55% of accident happen in Daylight</h1>

<h1>Question 2 how many accidents happen in accident severity?</h1>

In [41]:
acci_2020['Accident_Severity'].value_counts()

Accident_Severity
Slight     59875
Serious     9346
Fatal        942
Name: count, dtype: int64

In [52]:
acci_2020['Accident_Severity'].value_counts().sum()

np.int64(70163)

<h1> insight 2 = There are 85.33% of the accident falls under the "slight" category happen in 2020</h1>

<b>
<h1>Question 2 How many accidents occurred in each district area?</h1>

In [48]:
acci_2020['District Area'].value_counts()

District Area
Birmingham                   1438
Leeds                         921
Manchester                    721
Glasgow City                  676
Sheffield                     600
                             ... 
County Durham                   0
Shropshire                      0
Cheshire West and Chester       0
Cornwall                        0
Central Bedfordshire            0
Name: count, Length: 422, dtype: int64

<h1>insight 3 = most frequent accident happen in Birmingham district</h1>

<b>
<b>
<b>
<b>
<h1> Question 4. What vehicle type have the most incident?</h1>

In [53]:
acci_2020["Vehicle_Type"].value_counts()

Vehicle_Type
Car                                      52368
Van / Goods 3.5 tonnes mgw or under       3799
Motorcycle over 500cc                     2823
Bus or coach (17 or more pass seats)      2735
Goods 7.5 tonnes mgw and over             1892
Motorcycle 125cc and under                1728
Taxi/Private hire car                     1485
Motorcycle over 125cc and up to 500cc      812
Motorcycle 50cc and under                  745
Goods over 3.5t. and under 7.5t            664
Other vehicle                              622
Agricultural vehicle                       228
Minibus (8 - 16 passenger seats)           224
Pedal cycle                                 34
Data missing or out of range                 4
Ridden horse                                 0
Name: count, dtype: int64

<h1>Insight 4. 74.63% of the accident is a car accidents</h1>

<h1> BiVariate</h1>

In [58]:
acci_2021

Index Accident_Severity Accident Date   Latitude  \
352706  200901BS70001           Serious    2021-01-01  51.512273   
352707  200901BS70002           Serious    2021-01-05  51.514399   
352708  200901BS70003            Slight    2021-01-04  51.486668   
352709  200901BS70004           Serious    2021-01-05  51.507804   
352710  200901BS70005           Serious    2021-01-06  51.482076   
...               ...               ...           ...        ...   
516241       2.01E+12            Slight    2021-11-12  55.358904   
516242       2.01E+12            Slight    2021-11-12  55.166976   
516247       2.01E+12            Slight    2021-12-01  55.115346   
516250       2.01E+12            Slight    2021-12-01  55.313433   
516253       2.01E+12            Slight    2021-12-09  55.115088   

              Light_Conditions           District Area  Longitude  \
352706                Daylight  Kensington and Chelsea  -0.201349   
352707                Daylight  Kensington and Chelsea  -0.199248   
352708                Daylight  Kensington and Chelsea  -0.179599   
352709                Daylight  Kensington and Chelsea  -0.203110   
352710   Darkness - lights lit  Kensington and Chelsea  -0.173445   
...                        ...                     ...        ...   
516241                Daylight   Dumfries and Galloway  -3.513667   
516242  Darkness - no lighting   Dumfries and Galloway  -3.389896   
516247  Darkness - no lighting   Dumfries and Galloway  -3.409426   
516250   Darkness - lights lit   Dumfries and Galloway  -3.454262   
516253  Darkness - no lighting   Dumfries and Galloway  -3.376962   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
352706                     1                   2                     Dry   
352707                    11                   2             Wet or damp   
352708                     1                   2                     Dry   
352709                     1                   2            Frost or ice   
352710                     1                   2                     Dry   
...                      ...                 ...                     ...   
516241                     1                   2             Wet or damp   
516242                     1                   1             Wet or damp   
516247                     1                   2            Frost or ice   
516250                     1                   1                    Snow   
516253                     2                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
352706      One way street               Urban     Fine no high winds   
352707  Single carriageway               Urban     Fine no high winds   
352708  Single carriageway               Urban     Fine no high winds   
352709  Single carriageway               Urban                  Other   
352710  Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
516241    Dual carriageway               Rural  Raining no high winds   
516242    Dual carriageway               Rural  Raining no high winds   
516247  Single carriageway               Rural                  Other   
516250    Dual carriageway               Rural  Snowing no high winds   
516253  Single carriageway               Rural     Fine no high winds   

                 Vehicle_Type    Year  Month   Day  DayOfWeek  
352706                    Car  2021.0    1.0   1.0        4.0  
352707  Taxi/Private hire car  2021.0    1.0   5.0        1.0  
352708  Taxi/Private hire car  2021.0    1.0   4.0        0.0  
352709  Motorcycle over 500cc  2021.0    1.0   5.0        1.0  
352710                    Car  2021.0    1.0   6.0        2.0  
...                       ...     ...    ...   ...        ...  
516241  Motorcycle over 500cc  2021.0   11.0  12.0        4.0  
516242                    Car  2021.0   11.0  12.0        4.0  
5162

<h1> Question 5. Is there a relationship between accident severity and weather conditions?</h1>

In [60]:
acci_2021['Accident_Severity'].value_counts()

Accident_Severity
Slight     56491
Serious     8858
Fatal        823
Name: count, dtype: int64

In [79]:
sev_weather = acci_2021.groupby(['Accident_Severity', 'Weather_Conditions']).size()
sev_weather.unstack()

Weather_Conditions  Fine + high winds  Fine no high winds  Fog or mist  Other  \
Accident_Severity                                                               
Fatal                              14                 651            8     19   
Serious                           105                7144           57    222   
Slight                            603               43368          333   1819   

Weather_Conditions  Raining + high winds  Raining no high winds  \
Accident_Severity                                                 
Fatal                                 15                     97   
Serious                              109                    997   
Slight                               694                   7434   

Weather_Conditions  Snowing + high winds  Snowing no high winds  
Accident_Severity                                                
Fatal                                  0                      7  
Serious                               17                     93  
Slight                               117                    905

<h1>insight 5. The majority of accidents (both slight and serious) happen in fine weather with no high winds (43,368 slight, 7,144 serious, 651 fatal).</h1>

<h1>Question 6 Which type of area (urban or rural) has the highest number of slight, serious, and fatal accidents?</h1>

In [82]:
acci_2021.groupby(['Accident_Severity', 'Urban_or_Rural_Area']).size()

Accident_Severity  Urban_or_Rural_Area
Fatal              Rural                    525
                   Unallocated                0
                   Urban                    298
Serious            Rural                   3825
                   Unallocated                0
                   Urban                   5033
Slight             Rural                  19573
                   Unallocated                0
                   Urban                  36918
dtype: int64

<h1>Insight 6. Urban areas have more accidents overall, but they tend to be less severe. Rural areas have fewer accidents, but they are more deadly due to higher speeds and possibly fewer safety measures.</h1>

<h1>question 7  Does accident severity vary based on different light conditions?</h1>

In [83]:
acci_2021

Index Accident_Severity Accident Date   Latitude  \
352706  200901BS70001           Serious    2021-01-01  51.512273   
352707  200901BS70002           Serious    2021-01-05  51.514399   
352708  200901BS70003            Slight    2021-01-04  51.486668   
352709  200901BS70004           Serious    2021-01-05  51.507804   
352710  200901BS70005           Serious    2021-01-06  51.482076   
...               ...               ...           ...        ...   
516241       2.01E+12            Slight    2021-11-12  55.358904   
516242       2.01E+12            Slight    2021-11-12  55.166976   
516247       2.01E+12            Slight    2021-12-01  55.115346   
516250       2.01E+12            Slight    2021-12-01  55.313433   
516253       2.01E+12            Slight    2021-12-09  55.115088   

              Light_Conditions           District Area  Longitude  \
352706                Daylight  Kensington and Chelsea  -0.201349   
352707                Daylight  Kensington and Chelsea  -0.199248   
352708                Daylight  Kensington and Chelsea  -0.179599   
352709                Daylight  Kensington and Chelsea  -0.203110   
352710   Darkness - lights lit  Kensington and Chelsea  -0.173445   
...                        ...                     ...        ...   
516241                Daylight   Dumfries and Galloway  -3.513667   
516242  Darkness - no lighting   Dumfries and Galloway  -3.389896   
516247  Darkness - no lighting   Dumfries and Galloway  -3.409426   
516250   Darkness - lights lit   Dumfries and Galloway  -3.454262   
516253  Darkness - no lighting   Dumfries and Galloway  -3.376962   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
352706                     1                   2                     Dry   
352707                    11                   2             Wet or damp   
352708                     1                   2                     Dry   
352709                     1                   2            Frost or ice   
352710                     1                   2                     Dry   
...                      ...                 ...                     ...   
516241                     1                   2             Wet or damp   
516242                     1                   1             Wet or damp   
516247                     1                   2            Frost or ice   
516250                     1                   1                    Snow   
516253                     2                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
352706      One way street               Urban     Fine no high winds   
352707  Single carriageway               Urban     Fine no high winds   
352708  Single carriageway               Urban     Fine no high winds   
352709  Single carriageway               Urban                  Other   
352710  Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
516241    Dual carriageway               Rural  Raining no high winds   
516242    Dual carriageway               Rural  Raining no high winds   
516247  Single carriageway               Rural                  Other   
516250    Dual carriageway               Rural  Snowing no high winds   
516253  Single carriageway               Rural     Fine no high winds   

                 Vehicle_Type    Year  Month   Day  DayOfWeek  
352706                    Car  2021.0    1.0   1.0        4.0  
352707  Taxi/Private hire car  2021.0    1.0   5.0        1.0  
352708  Taxi/Private hire car  2021.0    1.0   4.0        0.0  
352709  Motorcycle over 500cc  2021.0    1.0   5.0        1.0  
352710                    Car  2021.0    1.0   6.0        2.0  
...                       ...     ...    ...   ...        ...  
516241  Motorcycle over 500cc  2021.0   11.0  12.0        4.0  
516242                    Car  2021.0   11.0  12.0        4.0  
5162

In [84]:
severity_light = acci_2021.groupby(['Accident_Severity', 'Light_Conditions']).size()
print(severity_light)

Accident_Severity  Light_Conditions           
Fatal              Darkness - lighting unknown        6
                   Darkness - lights lit            184
                   Darkness - lights unlit            4
                   Darkness - no lighting           147
                   Daylight                         482
Serious            Darkness - lighting unknown       76
                   Darkness - lights lit           1904
                   Darkness - lights unlit           34
                   Darkness - no lighting           736
                   Daylight                        6108
Slight             Darkness - lighting unknown      561
                   Darkness - lights lit          11210
                   Darkness - lights unlit          225
                   Darkness - no lighting          2873
                   Daylight                       41622
dtype: int64


<h1>Insight 7. Well-lit areas (Darkness - lights lit) saw a large number of accidents, while No lighting areas had fewer but more severe accidents 147 fatal accidents</h1>

<h1>Question 8. Which district has the most accidents in well-lit darkness?</h1>

In [98]:
well_lit_darkness = acci_2021[acci_2021['Light_Conditions'] == 'Darkness - lights lit']

district_well_lit_accidents = well_lit_darkness.groupby('District Area').size().reset_index(name='Accident_Count')

district_well_lit_accidents = district_well_lit_accidents.sort_values(by='Accident_Count', ascending=False)

print(district_well_lit_accidents)

          District Area  Accident_Count
24           Birmingham             364
195               Leeds             204
208          Manchester             192
147        Glasgow City             179
36             Bradford             164
..                  ...             ...
291             Rutland               0
63       Castle Morpeth               0
58              Caradon               0
22   Berwick-upon-Tweed               0
401       West Somerset               0

[422 rows x 2 columns]


<h1>Insight 8. Birmingham has the highest number of accidents (364), Leeds (204) and Manchester (192) follow closely, Glasgow City (179) and Bradford (164), Some districts, like Rutland, Castle Morpeth, Caradon, and West Somerset, report zero accidents</h1>

<h1>Question 9. Do more severe accidents tend to have a higher number of casualties?</h1>

In [99]:
severity_casualty = acci_2021.groupby('Accident_Severity')['Number_of_Casualties'].sum()
severity_casualty

Accident_Severity
Fatal       1483
Serious    13002
Slight     74921
Name: Number_of_Casualties, dtype: int64

In [101]:
accidents_per_severity = acci_2021['Accident_Severity'].value_counts()

casualties_per_severity = acci_2021.groupby('Accident_Severity')['Number_of_Casualties'].sum()

avg_casualties_per_severity = casualties_per_severity / accidents_per_severity

avg_casualties_per_severity

Accident_Severity
Fatal      1.801944
Serious    1.467826
Slight     1.326247
dtype: float64

<h1>Insight 9. Fatal accidents have the highest average casualties per accident (1.80 casualties per accident), Serious accidents have an average of 1.47 casualties per accident, Slight accidents have the lowest average casualties per accident (1.32 casualties per accident). in conclusion, The severity of an accident correlates with the number of casualties. Fatal and serious accidents tend to involve more people, making them a greater concern for road safety interventions.</h1>

<h1>Question 10. Does the severity of accidents vary by month? Are certain months more prone to fatal or serious accidents?</h1>

In [102]:
severity_month = acci_2021.groupby(['Month', 'Accident_Severity']).size().unstack()
severity_month

Accident_Severity  Fatal  Serious  Slight
Month                                    
1.0                   66      582    3897
2.0                   61      588    4389
3.0                   68      663    4469
4.0                   84      731    4128
5.0                   63      763    4439
6.0                   62      763    4803
7.0                   68      791    4796
8.0                   79      757    4373
9.0                   63      831    4755
10.0                  70      802    5127
11.0                  67      765    5520
12.0                  72      822    5795

<h1>Insight 10. Serious accidents are generally higher from April (731) to December (822), Slight accidents make up the majority of incidents in every month (80%+ consistently). December, despite having the highest number of accidents, has a lower fatality rate (1.09%) compared to other months.</h1>

<h1>Multi Variate</h1>

<h1>Question . Do certain districts have higher accident severity during specific light conditions?</h1>

In [92]:
district_severity_light = acci_2021.groupby(['District Area', 'Accident_Severity', 'Light_Conditions']).size().reset_index(name='Count')
district_severity_light

District Area Accident_Severity             Light_Conditions  Count
0     Aberdeen City             Fatal  Darkness - lighting unknown      0
1     Aberdeen City             Fatal        Darkness - lights lit      0
2     Aberdeen City             Fatal      Darkness - lights unlit      0
3     Aberdeen City             Fatal       Darkness - no lighting      0
4     Aberdeen City             Fatal                     Daylight      0
...             ...               ...                          ...    ...
6325           York            Slight  Darkness - lighting unknown      1
6326           York            Slight        Darkness - lights lit     39
6327           York            Slight      Darkness - lights unlit      3
6328           York            Slight       Darkness - no lighting      9
6329           York            Slight                     Daylight    135

[6330 rows x 4 columns]

<h1>Insight . Urban districts may have more accidents in well-lit conditions. Due to better street lighting, urban areas might have more slight accidents at night compared to rural areas. Rural districts may have more severe accidents in darkness.</h1>

In [64]:

f_stats, p_value = f_oneway(acci_2021[acci_2021['Urban_or_Rural_Area'] == 'Urban']['Number_of_Casualties'],
                            acci_2021[acci_2021['Urban_or_Rural_Area'] == 'Rural']['Number_of_Casualties'])

In [65]:
p_value

np.float64(3.987975195484272e-158)